Yank/restraints.py

```diff
@@ -1564,46 +1737,73 @@ def _pick_restrained_atoms(self, sampler_state, topography):
         Future updates can further refine this algorithm.
 
         """
-        t = md.Trajectory(sampler_state.positions / unit.nanometers, topography.topology)
+        # No need to determine parameters if atoms have been given.
+        if self.restrained_receptor_atoms is not None and self.restrained_ligand_atoms is not None:
+            return self.restrained_receptor_atoms + self.restrained_ligand_atoms
+
+        # If receptor and ligand atoms are explicitly provided, use those.
+        heavy_ligand_atoms = self.restrained_ligand_atoms
+        heavy_receptor_atoms = self.restrained_receptor_atoms
+
+        # Otherwise we restrain only heavy atoms.
+        heavy_atoms = set(topography.topology.select('not element H').tolist())
+        # Intersect heavy atoms with receptor/ligand atoms (s1&s2 is intersect).
+        if heavy_ligand_atoms is None:
+            heavy_ligand_atoms = set(topography.ligand_atoms) & heavy_atoms
+        if heavy_receptor_atoms is None:
+            heavy_receptor_atoms = set(topography.receptor_atoms) & heavy_atoms
 
-        # Determine heavy atoms. Using sets since lists should be unique anyways
-        heavy_atoms = set(topography.topology.select('not element H'))
-
-        # Intersect heavy atoms with receptor/ligand atoms (s1&s2 is intersect)
-        heavy_ligand_atoms = set(topography.ligand_atoms) & heavy_atoms
-        heavy_receptor_atoms = set(topography.receptor_atoms) & heavy_atoms
         if len(heavy_receptor_atoms) < 3 or len(heavy_ligand_atoms) < 3:
             raise ValueError('There must be at least three heavy atoms in receptor_atoms '
                              '(# heavy {}) and ligand_atoms (# heavy {}).'.format(
                                      len(heavy_receptor_atoms), len(heavy_ligand_atoms)))
 
-        # Find valid pairs of ligand/receptor atoms within a cutoff
+        # If r3 or l1 atoms are given. We have to pick those.
+        if isinstance(heavy_receptor_atoms, list):
```

Comment:
Shouldn't be a problem. `heavy_receptor_atoms` and `heavy_ligand_atoms` will be `list`s only if the user has specified them, otherwise they'll be `set`s.